### PR TITLE
Add flag category support (#796)

### LINKED ISSUE #259: Flag Groups
It might be nice to be able to have `command.Flags` be a `[][]Flag` for the purposes of being able to group flags in the command help output.
In addition instead of a straight up `[][]Flag`, a new `FlagGroup` struct type may be better, where we can define a name/other fields.

WDYT?

### PR DESCRIPTION
**Pull request by meatballhat**
## What type of PR is this?

- feature

## What this PR does / why we need it:

See #796 

## Which issue(s) this PR fixes:

Closes #259 

## Special notes for your reviewer:

This PR is a best-effort porting of #796 so detailed criticism is very much appreciated 🙇🏼 

## Release Notes

```release-note
add flag category support
```

### CODE REVIEW

File: category.go [hunk 1]
```diff
@@ -77,3 +77,48 @@ func (c *commandCategory) VisibleCommands() []*Command {
 	}
 	return ret
 }
+
+// FlagCategories is a slice of *FlagCategory.
+type FlagCategories []*FlagCategory
+
+// FlagCategory is a category containing commands.
```
Comment (dearchap) [inline category.go:84]: containing flags

File: category.go [hunk 2]
```diff
@@ -77,3 +77,48 @@ func (c *commandCategory) VisibleCommands() []*Command {
 	}
 	return ret
 }
+
+// FlagCategories is a slice of *FlagCategory.
+type FlagCategories []*FlagCategory
```
Comment (dearchap) [inline category.go:82]: map[category]Flag is better ?

**Review by dearchap — DISMISSED** [review]

File: category.go (hunk 2 above)
Comment (meatballhat) [reply category.go:82]: Agreed!

Comment (meatballhat) [reply category.go:82]: Welp, I started trying to switch the implementation to `map[string][]Flag`, and it seems non-trivial to me so far... but I'm also nervous about introducing `FlagCategories` and `FlagCategory` and being committed to their behaviors for the rest of the `v2` series 😐

File: flag.go [hunk 3]
```diff
@@ -132,6 +132,14 @@ type VisibleFlag interface {
 	IsVisible() bool
 }
 
+// CategorizableFlag is an interface that allows us to potentially
+// use a flag in a categorized representation.
+type CategorizableFlag interface {
+	VisibleFlag
+
+	GetCategory() string
```
Comment (dearchap) [inline flag.go:140]: Rename to just Category(). golang doesnt recommend using GetXXX

File: app.go [hunk 4]
```diff
@@ -181,6 +183,15 @@ func (a *App) Setup() {
 		if c.HelpName == "" {
 			c.HelpName = fmt.Sprintf("%s %s", a.HelpName, c.Name)
 		}
+
+		fc := newFlagCategories()
```
Comment (dearchap) [inline app.go:187]: Move this out into a separate function, with a []Flag as argument

File: app.go [hunk 5]
```diff
@@ -481,6 +499,14 @@ func (a *App) VisibleCommands() []*Command {
 	return ret
 }
 
+// VisibleFlagCategories returns a slice containing all the categories with the flags they contain
+func (a *App) VisibleFlagCategories() []VisibleFlagCategory {
+	if a.flagCategories == nil {
+		a.flagCategories = newFlagCategories()
```
Comment (dearchap) [inline app.go:505]: cant you return nil or [] if a.flagCategories is nil ? Its not nice behaviour to update flagCategories here

File: category.go [hunk 6]
```diff
@@ -77,3 +79,82 @@ func (c *commandCategory) VisibleCommands() []*Command {
...
+
+func (f *defaultFlagCategories) AddFlag(category string, fl Flag) {
+	if _, ok := f.m[category]; !ok {
+		f.m[category] = &defaultVisibleFlagCategory{name: category, m: map[string]Flag{}}
+	}
+
+	f.m[category].m[fl.String()] = fl
+}
+
+func (f *defaultFlagCategories) VisibleCategories() []VisibleFlagCategory {
```
Comment (dearchap) [inline category.go:109]: where is the visibility of the category being checked here ?

File: category.go [hunk 7]
```diff
@@ -77,3 +79,82 @@ func (c *commandCategory) VisibleCommands() []*Command {
...
+	catNames := []string{}
+	for name := range f.m {
+		catNames = append(catNames, name)
+	}
+
+	sort.Strings(catNames)
```
Comment (dearchap) [inline category.go:115]: isnt f.m already sorted ?

File: command.go [hunk 8]
```diff
@@ -280,6 +281,14 @@ func (c *Command) startApp(ctx *Context) error {
 	return app.RunAsSubcommand(ctx)
 }
 
+// VisibleFlagCategories returns a slice containing all the visible flag categories with the flags they contain
+func (c *Command) VisibleFlagCategories() []VisibleFlagCategory {
+	if c.flagCategories == nil {
```
Comment (dearchap) [inline command.go:286]: not a good idea to initialize c.flagCategories here .

File: command.go [hunk 9]
```diff
@@ -280,6 +281,14 @@ func (c *Command) startApp(ctx *Context) error {
...
+		c.flagCategories = newFlagCategories()
```
Comment (dearchap) [inline command.go:287]: return nil or empty slice here ?

File: flag_bool.go [hunk 10]
```diff
@@ -19,6 +19,7 @@ type BoolFlag struct {
 	DefaultText string
 	Destination *bool
 	HasBeenSet  bool
+	Category    string
```
Comment (dearchap) [inline flag_bool.go:22]: @meatballhat How would this tie-in with your generator code ?

File: flag.go (hunk 3 above)
Comment (meatballhat) [reply flag.go:140]: What about the conflict with the field name `Category`? 🤔 I agree that `GetXXX` is generally bad style, but this is a language-level wart we're working around.

File: flag_bool.go (hunk 10 above)
Comment (meatballhat) [reply flag_bool.go:22]: If all the flag types defined in this repo have `Category` fields, then it'd be part of the struct template. Or did you mean the potential merge conflict? (or something else?)

File: app.go (hunk 4 above)
Comment (meatballhat) [reply app.go:187]: Excellent 👍🏼

File: app.go (hunk 5 above)
Comment (meatballhat) [reply app.go:505]: I agree!

File: category.go (hunk 7 above)
Comment (meatballhat) [reply category.go:115]: IIUC ranging over a `map` has undefined sort order: <https://go.dev/play/p/aR7JKg4ae8r> Or did you mean something else? 🤔

File: command.go (hunk 8 above)
Comment (meatballhat) [reply command.go:286]: I agree!

**Pull request merged**